src/workspaces/useWorkspaces.composable.ts

```diff
@@ -0,0 +1,75 @@
+import { withHandlers } from '@terra-ui-packages/core-utils';
+import { useState } from 'react';
+import { FieldsArg, WorkspaceDataProvider } from 'src/libs/ajax/workspaces/providers/WorkspaceProvider';
+import { withErrorReporting } from 'src/libs/error';
+import { useCancellation, useOnMount } from 'src/libs/react-utils';
+import { withBusyState } from 'src/libs/utils';
+import { WorkspaceWrapper } from 'src/libs/workspace-utils';
+
+export interface UseWorkspacesStateResult {
+  workspaces: WorkspaceWrapper[];
+  refresh: () => Promise<void>;
+  loading: boolean;
+}
+
+export type UseWorkspacesState = (
+  fields?: FieldsArg,
+  stringAttributeMaxLength?: string | number
+) => UseWorkspacesStateResult;
+
+const defaultFieldsArgs: FieldsArg = [
+  'accessLevel',
+  'public',
+  'workspace',
+  'workspace.state',
+  'workspace.attributes.description',
+  'workspace.attributes.tag:tags',
+  'workspace.workspaceVersion',
+];
+
+export type WorkspaceDataProviderNeeds = Pick<WorkspaceDataProvider, 'list'>;
+export interface UseWorkspacesDeps {
+  /**
+   * Provides the data accessor for workspaces list
+   */
+  workspaceProvider: WorkspaceDataProviderNeeds;
+
+  /**
+   * Provides state retention of current workspaces list.  Implementation can just return useState
+   * or an equivalent like useSettableStore.  Implementation is expected to be a React Hook that will
+   * re-render a consuming component on state update.
+   */
+  useWorkspacesState: () => [WorkspaceWrapper[], (newValue: WorkspaceWrapper[]) => void];
```

Comment:
I'm not keen on having this injected. I think `useWorkspaces` vs `useWorkspacesState` is a subtle difference and likely to be confusing. I found it difficult to understand when reviewing this PR.

To support Terra UI's global workspaces store, I think I'd propose an optional `onLoad` (or similar name) callback similar to the `onError` callback will have to be added to support error notifications.

Or... Terra UI's `useWorkspaces` could wrap the base `useWorkspaces` and use an effect to update the store.
```ts
const useWorkspaces = () => {
  const result = baseUseWorkspaces();
  useEffect(() => {
    workspacesStore.set(result.workspaces);
  }, [result.workspaces])
  return result
})
```

That would simplify the `makeUseWorkspaces` interface.

Reply:
Ideally, I suppose it would be preferable to take an initial value. However, practically, it doesn't matter. The workspaces store is initialized with an empty array and since `useWorkspacesComposable` is run at module import/app load, the initial value will always be an empty array. We'd only have to worry about this if we started caching the workspaces store in local storage, etc.

Reply:
Oh, I misunderstood...

But I still think that returning an initial value from the store can be done without injecting the state into the generic useWorkspaces.

```ts
const useWorkspaces = (...args) => {
  const { workspaces, loading, error } = baseUseWorkspaces(...args);
  useEffect(() => {
    workspacesStore.set(workspaces);
  }, [workspaces])

  const storedWorkspaces = useStore(workspacesStore);
  return { workspaces: storedWorkspaces, loading, error };
})
```

And the same pattern could probably be used for error notifications, allowing the generic useWorkspaces to have the same interface as the Terra UI specific useWorkspaces:

```ts
const useWorkspaces = (...args) => {
  const { workspaces, loading, error } = baseUseWorkspaces(...args);
  
useEffect(() => {
    workspacesStore.set(workspaces);
  }, [workspaces])

  useEffect(() => {
    if (error) {
       reportError('Error loading workspaces' error);
    }
  }, [error])

  const storedWorkspaces = useStore(workspacesStore);
  return { workspaces: storedWorkspaces, loading, error };
})
```

Reply:
hmmm....

Reply:
ok, what you suggest will work... as long as useWorkspaces is the only source of truth.... the way the atoms are intended to function is a more centralized store which many parts of the UI can treat as the source of truth.  Looking through the suggestion and the hook internals of useWorkspaces, it would have to assume an empty [] array as initial value on mount... clearing out any previously set value in workspacesStore.

Reply:
> Looking through the suggestion and the hook internals of useWorkspaces, it would have to assume an empty [] array as initial value on mount... clearing out any previously set value in workspacesStore.

Ah, because this would reset the store with the initial workspaces list returned from `baseUseWorkspaces`?

```ts
const { workspaces, loading, error } = baseUseWorkspaces(workspacesProvider, ...args)
  
useEffect(() => {
  workspacesStore.set(workspaces)
}, [workspaces])
```

I think that could fixed by only updating the store when loading is done:

```ts
const { workspaces, loading, error } = baseUseWorkspaces(workspacesProvider, ...args)
  
useEffect(() => {
  if (!loading) {
    workspacesStore.set(workspaces)
  }
}, [workspaces, loading])
```

Reply:
nope, useWorkspaces onMount would still clear out central store to [] empty array since loading is false then too.   We would likely have to pass in an initial value, which creates it's own spagetti.... landing back to my current proposal of just providing a store dependency which feels cleaner.

Reply:
I think that's an issue with the initial value of `loading` then.

Reply:
baby steps.... I am resisting bringing in useLoadedData to keep the PR risk down.... useLoadedData initial NoneState could actually help here if you want to keep the wrapper-hook-adds-new-concern (store) concern.... but I argue that the store dependency handles this well....

Because if we want to preserve the current assumption that atoms are CENTAL sources of truth, making the store a dependency is cleaner in my opinion then the wrapper cruft required... what about the case where an external actor changes the store?  useLoadedData needs that update to hold up on the current set of assumptions...

Reply:
My objection to the store as dependency is that it allows Terra UI specific concerns to leak into shared code.

> what about the case where an external actor changes the store? useLoadedData needs that update to hold up on the current set of assumptions...

A `useWorkspaces` hook that uses the workspace store and wraps `useLoadedData` would handle this.

Reply:
hmm... ok, after our chat, I started to kick this around locally to see the benefits of going with useLoadedData variant called useAutoLoadedData that we talked about... I will have to pick this up next week after my mini-vacation:

```
export const useWorkspaces = (
  fieldsArg?: FieldsArg,
  stringAttributeMaxLength?: number
): UseWorkspacesStateResult => {

  const signal = useCancellation();
  const getData = async (): Promise<WorkspaceWrapper[]> => await workspaceProvider.list(fields, {stringAttributeMaxLength, signal});

  const [ workspaces, updateWorkspaces] = useAutoLoadedData<WorkspaceWrapper[]>(getData, {
    onError: () => { reportError('Error loading workspace list'); }
  });
  useEffect(() => {
    if (workspaces.status == 'Ready') {
      workspacesStore.set(workspaces.state);
    }
  }, [workspaces]);
  
  const fields: FieldsArg = fieldsArg || defaultFieldsArgs;
  
  return { 
    workspaces: workspacesStore.get(), 
    refresh: () => updateWorkspaces(getData), 
    loading: workspaces.status == 'Loading'
  };
};

```